rfcs/0000-execute-builds-in-standard-containers.md

```diff
@@ -0,0 +1,132 @@
+       will exit with the same error code, forcing the pod to enter a failed state. There is also a configurable
+       timeout period that will specify how long the container is going to wait for th eprevious steps to complete
+       before exiting with an error. Note: For `prepare` we can provide an empty `wait step` so that it begins
+       executing immediately.
+
+   Example of possible `build-waiter` implementation is in the [spike][kpack-spike-build-waiter].
+
+2. Convert the existing init containers in `build_pod.go` to be regular containers and add an init container to copy
+   the `build-waiter` to a shared mount
+3. Using a similar [step modifier][step-modifier] approach , we can put this new behavior behind a feature gate and
+   only execute in regular containers if enabled by the user. We can do this at the kpack controller level or at the
+   image level via a spec change or annotation. In the event that the feature flag is not set, kpack will continue
+   to execute builds in init containers.
+4. Due to the fact that sidecars such as istio-proxy, do not exit when all other containers exit, we will need to add a
+   method for stopping the proxy once a build exits
+    1. To do this we can update the running pod's `spec.activeDeadlineSeconds` to 1 which will cause the pod to go into
+       a failed state and kill all containers after 1 second. This will happen in the case that during a build
+       reconcile, kpack detects that a pod is running, but completion container is not. (The completion container will
+       exit in the event that another build container fails)
+    2. This will kill the sidecar, but it will also leave our pod in a failed state with the message `DeadlineExceeded`,
+       leading the reconciler to believe
+       the build has failed. To circumvent this we will add a check that if a pod has failed with the
+       reason `DeadlineExceeded` AND the completion container has successfully completed, then the build is actually
+       successful.
+
+[kpack-spike-build-waiter]: https://github.com/pivotal/kpack/blob/execute-builds-in-regular-containers/cmd/build-waiter/main.go
+
+[step-modifier]: https://github.com/pivotal/kpack/blob/44257a70c2c9e9ac5703eadc871c7c9be22cfadc/pkg/apis/build/v1alpha2/build_pod.go#L181
+
+## Complexity
+
+* Our current build logging relies heavily on the fact that our builds exist in init containers so that logic will
+  need to be reworked to support this change. This logging change will also need to be aware of any feature gates
+  to enable this feature
+* We would need to create a build-waiter image for windows builds as well if windows support is needed
+
+## Prior Art
+
+### 1. tekton
+
+https://github.com/tektoncd/pipeline/tree/main/cmd/entrypoint#waiting-for-sidecars
+
+### 2. kpack spike
+
+https://github.com/pivotal/kpack/pull/1019
+
+## Alternatives
+
+* We can create tekton [`TaskRuns`][tekton-task-run] to execute each step of the build since tekton has come up with
```

Comment:
I think due to the maintenance benefits of this alternative, I would vote for it

Reply:
I think the maintenance burden will be similar in either case because the build waiter image will likely be fairly static so the bulk of the maintenance will be in the build pod generation and reconciliation logic which will exist for both the main proposed solution and the alternative